feature(nix): Add the ability to build under NixOS and from under Nix

fisuri opened the pull request.

fisuri committed:
feat(nix): add build

fisuri committed:
chore: add nix instruction

fisuri committed:
chore: add endings for a code block

fisuri committed:
chore: install a link to the original repository

MMatveyA commented:
# Общие замечания

- Ты вообще никак не задействуешь `cmake`, не указываешь ни тип сборки (`release`), ни количество ядер для сборки

# `default.nix`

- `qt6.full` слишком тяжёлый. Лучше использовать `qt6.qtbase` и другие необходимые модули точечно.
- Нет `installPhase`. Пример (лучше задать путь `build/src/dnd-vtt` в виде переменной):
    
        installPhase = ''
          mkdir -p $out/bin
          cp build/src/dnd-vtt $out/bin/
        '';
- Нет тестов через `checkPhase`. Хоть сейчас в самом коде их и нет, но позже они появляется (я буду использовать `ctest`, что можно увидеть в GH Actions).
- `name = "dnd-vtt";` нельзя просто так использовать, так как необходимо указывать `pname + version`.
- Ты не используешь `wrapQtAppsHook`. Без него нет чёткой уверенности, что компилятор найдёт и включит в исполняемый файл все qml файлы.
- Правильнее использовать `lib.license.lgpl3` и `lib.platforms.linux`.
- Дублирование `qt6.full` в `nativeBuildInputs` и `buildInputs` — избыточно.
- Нельзя указывать `src` как `./.`. Для этого специально есть `self`.

# `flake.nix`

- В проектах НЕЛЬЗЯ использовать `nixpkgs-unstable` иначе вся детерминированность моментально падает. Сейчас актуальна `nixos-25.05`, но скоро выйдет `nixos-25.11`.
- В поле `description` осталось шаблонная строчка, точно не имеющая отношение к проекту.
- Ты не добавил возможность запускать приложение через `nix run`, что-то типа

```nix
    apps = {
        dnd-vtt = {
            type = "app";
            program = "${self.packages.${system}.dnd-vtt}/bin/dnd-vtt";
        };
    };
```

# `shell.nix`

- В `mkShell` должен быть `buildInputs`, а не `nativeBuildInputs`.
- `LD_LIBRARY_PATH` это костыль. В [документации](https://nixos.wiki/wiki/Qt) чётко написано, что надо использовать `wrapQtAppsHook`.

fisuri commented:
> * Ты вообще никак не задействуешь `cmake`, не указываешь ни тип сборки (`release`), ни количество ядер для сборки

cmake нужен, иначе не соберется, по умолчанию собирается в релиз

> * Нет `installPhase`. Пример (лучше задать путь `build/src/dnd-vtt` в виде переменной):

он и не нужен, nix автоматический определяет

fisuri commented:
> * Ты вообще никак не задействуешь `cmake`, не указываешь ни тип сборки (`release`), ни количество ядер для сборки

cmake нужен, иначе не соберется, по умолчанию собирается в релиз, количество ядер он сам определяет

> * Нет `installPhase`. Пример (лучше задать путь `build/src/dnd-vtt` в виде переменной):

он и не нужен, nix автоматический определяет
```bash
dr-xr-xr-x   - root  1 янв  1970  result
dr-xr-xr-x   - root  1 янв  1970 └──  bin
.r-xr-xr-x 43k root  1 янв  1970     ├──  .dnd-vtt-wrapped
.r-xr-xr-x 16k root  1 янв  1970     └──  dnd-vtt
```

> * Ты не добавил возможность запускать приложение через `nix run`, что-то типа

`nix run` запускает дефолтное приложение
```nix
packages.default = pkgs.callPackage ./default.nix {inherit self;};
```

> * `LD_LIBRARY_PATH` это костыль. В [документации](https://nixos.wiki/wiki/Qt) чётко написано, что надо использовать `wrapQtAppsHook`.

это разные вещи

fisuri committed:
chore: meeting the requirements

chore: meeting the requirements

chore: meeting the requirements

chore: meeting the requirements

fisuri commented:
> * Нет тестов через `checkPhase`. Хоть сейчас в самом коде их и нет, но позже они появляется (я буду использовать `ctest`, что можно увидеть в GH Actions).

когда будут тесты, nix будет их выполнять сам

fisuri commented:
> * [документации](https://nixos.wiki/wiki/Qt)

используй wiki.nixos.org

MMatveyA commented:
> `nix run` запускает дефолтное приложение

Он плохо определяет его. Если собирать под MacOS через nix, то там образуется другое файловое дерево и испоняемый файл лежит по пути `result/dnd-vtt.app/Contents/MacOS/dnd-vtt`.

> это разные вещи

`wrapQtAppsHook` это набор предопределений, включающий в себя `LD_LIBRARY_PATH`.

> когда будут тесты, nix будет их выполнять сам

Он их отключает флагом `-DBUILD_TESTING=OFF`

> используй wiki.nixos.org

Нет такого веб-сайта. `curl` на запрос на этот адрес выдаёт

```text
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

А теперь по коду. Блок nativeBuildInputs  определяет зависимости, необходимые для сборки. В нашем случае это cmake. `buildInputs` определяет зависимости не только для сборки, но и для выполнения. Так как ты указал только блок `nativeBuildInputs`, то пользователю просто не установится Qt.

fisuri commented:
> Нет такого веб-сайта. `curl` на запрос на этот адрес выдаёт

https://wiki.nixos.org/wiki/Qt

fisuri commented:
> `LD_LIBRARY_PATH`

ты хочешь чтобы я это убрал?

fisuri committed:
feat: remove LD_LIBRARY_PATH

fisuri committed:
chore: meeting the requirements

fisuri committed:
feat: add direnv support
